### PR TITLE
[Release-1.6] Infer headless service tls mode

### PR DESCRIPTION
**Pull request by hzxuzhonghu**
manually cherry-pick #27105

Fixes: #28521

### CODE REVIEW

**Review by howardjohn — COMMENTED** [review]
Is this accurate
1.6 headless -> no auto mtls
1.7 headless -> auto mtls (with the exception it fails with "mixed" pods, but that is rare)
1.6 headless with this pr -> auto mtls (with the exception it fails with "mixed" pods, but that is rare)

?

My main concern is we keep changing the behavior of 1.6 in patch releases

**Comment by hzxuzhonghu** [comment]
Yes. 
>My main concern is we keep changing the behavior of 1.6 in patch releases

Yeah, we have changed many times. Hope this will be the last one.

**Comment by jacob-delgado** [comment]
cc @hzxuzhonghu with release-1.6 closing next week will you be able to get this done or should we close this out?

**Comment by hzxuzhonghu** [comment]
@jacob-delgado Any idea of the ci failure, it seems unrelated

**Comment by hzxuzhonghu** [comment]
rebased

**Comment by jacob-delgado** [comment]
cc @howardjohn thoughts on this before I approve it? It seems you had some concerns.

File: pilot/pkg/model/push_context.go [hunk 1]
```diff
@@ -1708,12 +1708,26 @@ func (ps *PushContext) BestEffortInferServiceMTLSMode(service *Service, port *Po
 		return MTLSUnknown
 	}
 
-	// First , check mTLS settings from beta policy (i.e PeerAuthentication) at namespace / mesh level.
+	// 1. Check service instances' tls mode, mainly used for headless service.
+	if service.Resolution == Passthrough {
+		instances, _ := ps.InstancesByPort(service, port.Port, nil)
```
Comment (howardjohn) [inline pilot/pkg/model/push_context.go:1713]: we had concerns about the performance of this in the original PR and made some (major) changes to make this more performant.

**Review by howardjohn — COMMENTED** [review]
to be honest, I am pretty concerned about this given we have changed the  behavior of auto mtls+headless multiple times _already_ on 1.6 causing patch release upgrade breakages, and this is a less performant version of the original pr

**Comment by hzxuzhonghu** [comment]
Can we close this, as we have already documented the right usage guide?

**Comment by jacob-delgado** [comment]
> Can we close this, as we have already documented the right usage guide?

Closing. Thanks @howardjohn and @hzxuzhonghu for addressing this.

**Pull request closed without merging**